### PR TITLE
perf(core): add ability to provide a test setup function to benchmarks

### PR DESCRIPTION
**Pull request by atscott**
This commit adds the ability to provide a test setup function to a micro benchmark. With this,
benchmarks can now handle expensive setup operations that may mutate state which needs
to be reset on each iteration. These setups will not be included in profile timings.

I modified this a little from what I had done in #37710 so that the time to run the test setup is tracked separately. If you think modifying the timestamp directly is more understandable (see changes in #37710), I can change it back.

### CODE REVIEW

File: packages/core/test/render3/perf/micro_bench.ts [hunk 1]
```diff
@@ -28,17 +28,25 @@ export interface Profile {
...
+ *     Because tests may rely on mutating some state, there may be a need to reset that state in
+ *     each iteration. In addition, the time to run the test setup is excluded from the performance
+ *     timing.
+ */
+export function createBenchmark(benchmarkName: string, testSetup?: () => void): Benchmark {
   const profiles: Profile[] = [];
 
   const benchmark = function Benchmark(profileName: string): Profile {
-    let iterationCounter: number = 0;
-    let timestamp: number = 0;
```
Comment (pkozlowski-opensource) [inline packages/core/test/render3/perf/micro_bench.ts:36]: Thnx for removing `number` type 👍

File: packages/core/test/render3/perf/micro_bench.ts [hunk 2]
```diff
@@ -28,17 +28,25 @@ export interface Profile {
...
 
-export function createBenchmark(benchmarkName: string): Benchmark {
+/**
+ * @param benchmarkName The name for this benchmark, used in the console.log for identification.
+ * @param testSetup A function that can be used to do the necessary set up for each iteration.
+ *     Because tests may rely on mutating some state, there may be a need to reset that state in
+ *     each iteration. In addition, the time to run the test setup is excluded from the performance
+ *     timing.
+ */
+export function createBenchmark(benchmarkName: string, testSetup?: () => void): Benchmark {
```
Comment (pkozlowski-opensource) [inline packages/core/test/render3/perf/micro_bench.ts:38]: nit: maybe rename `testSetup` => `beforeEach` as this plays the same role as Jasmine's  `beforeEach`, right?

Generally speaking we start to have a bit of confusion here since there are 2 levels (benchmark, profile) and only one function so it is not clear from the name description if this function runs once before all scenarios (profile) or once for a scenario (that is, if the number of runs is noOfIterations or noOfIterations*numberOfProfiles.

Not sure if this is a documentation or naming issue - what do you think?

**Review by pkozlowski-opensource — COMMENTED** [review]
@atscott I left a couple of comments but more importantly I would love to chat with you about this bench overall - I must say that I find it harder and harder to follow which isn't great since I wanted to keep it dead-simple (so it is "obvious" that nothing funky is going on - TBH I've seen way too much broken perf harness libraries...).

Let's chat offline to see if we can improve things (and if there is anything to improve here in the first place...).

**Review by pkozlowski-opensource — APPROVED** [review]
LGTM, thnx for doing the renames.

As discussed yesterday I think that we should simplify the micro-benchmark as it starts to be hard to follow. We can do so in a separate PR, though.

CI is failing atm so it will need cleanup.

File: packages/core/test/render3/perf/micro_bench.ts [hunk 3]
```diff
@@ -69,9 +78,20 @@ export function createBenchmark(benchmarkName: string): Benchmark {
...
+        if (beforeEachIteration) beforeEachIteration();
+        sampleStartTime = performance.now();
+        timeToRunTestSetup = 0;
         return runAgain;
       } else {
+        // If there is a test setup function provided, we need to run it before starting the next
+        // iteration.
+        if (beforeEachIteration) {
+          const start = performance.now();
+          beforeEachIteration();
```
Comment (mhevery) [inline packages/core/test/render3/perf/micro_bench.ts:92]: Imagine we have 
```
function underProfile1() {
  return 1 + 2;
}
function underProfile2() {
  return 2 + 1;
}
```

We could pass `underProfile1` to `createBenchmark` something like this:
```
const profile1 = createBenchmark(underProfile, ...);
profile1();
const profile2 = createBenchmark(underProfile, ...);
profile2();

```
But if we did the above the VM would inline first `underProfile1` and run it fast. Then it would get to `underProfile2` and VM would realize that the inlining was wrong and it would deopt. Going forward it would call `underProfile2` rather than inline. This would make it seems that `underProfile2` is slower than `underProfile1` when in reality they are the same.

For the above reason we do this instead:
```
const profile1 = createBenchmark(...);
while(profile1()) {
  1+2;
}
const profile2 = createBenchmark(...);
while(profile1()) {
  2+1;
}
```
Now instead of inlining `underProfile1` into `createBanchmark` the reverse is happening. (`createBanchmark` is inlined into `underProfile1` test) Because of this no deopt will happen and both tests will run same speed.

----

Your change does exactly the above. It causes the `beforeEach` to be inlined into the `createBanchmark` causing a deopt on second invocation. This is a problem.


----

But there is a second problem here. To subtract out cost of `beforeEachIteration` you use `performance.now()` which will only work if `beforeEachIteration` is significantly more expensive than resolution of `performance.now()`. In most cases it will not be which will introduce noise.

This algorithm specifically is designed to be able to measure things which are smaller than resolution of `performance.now()` by iterating many times over the work until we have work larger than `performance.now()` resolution and then dividing by number of iterations. Trying to subtract out `beforeEachIteration` will be futile in that case.

**Review by mhevery — CHANGES_REQUESTED** [review]
Sorry this approach introduces two issues described in more detail in comments:
1. The invocation of `beforeEach` will cause a deopt leading to wrong measurements.
2. The `performance.now()` timer resolution is not good enough to subtract out `beforeEach` 

I don't really know of a good way to solve the two issues.

**Comment by atscott** [comment]
To summarize after our discussion: 

I think we both agreed that the problematic part isn't the deopt, since the deopt happens with the `beforeEachIteration` and the change attempts to subtract the time it takes to run that from the benchmark (so it doesn't matter if it's fast/slow/different on each iteration). However, the real problem is the `performance.now()` resolution - we can't accurately subtract the `beforeEachIteration` runtime because (1) the time to resolve `performance.now` may be significant if the code being benchmarked is faster, and (2) `performance.now` intentionally has some "fuzziness" built in so again, if the benchmark code is very quick, this fuzziness could significantly affect results.

Closing. As a follow up, I need to try rewriting the `ngFor`/`watchCollection` benchmarks to not use the testSetup
